Kuadrant/kuadrant-operator
# Kuadrant Merge

This PR includes every single step, introduced by previous PRs, accomplishing the merge of the Kuadrant Controller repo and it's features within the Kuadrant Operator.
 
Closes https://github.com/Kuadrant/kuadrant-operator/issues/65
Closes https://github.com/Kuadrant/kuadrant-operator/issues/66

## Verification Steps
Every user guide should work as expected. Any reviewer should be able to reproduce them:

* [Simple rate limit for API owners](https://github.com/Kuadrant/kuadrant-operator/blob/582e038540395be8993e05059a26a77867a8cafa/doc/user-guides/simple-rl-for-api-owners.md)
* [Gateway rate limit for cluster operators](https://github.com/Kuadrant/kuadrant-operator/blob/582e038540395be8993e05059a26a77867a8cafa/doc/user-guides/gateway-rl-for-cluster-operators.md)
* And our best seller [Authenticated rate limit for API Owners](https://github.com/Kuadrant/kuadrant-operator/blob/582e038540395be8993e05059a26a77867a8cafa/doc/user-guides/authenticated-rl-for-api-owners.md)

Also, you could try this simple rate limit example:

1. Setup a local Kuadrant cluster
```bash
make local-setup
```

2. Apply Kuadrant object in order to deploy required dependencies (this will change in the future most probably)

```bash
kubectl -n kuadrant-system apply -f - <<EOF
---
apiVersion: kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  name: kuadrant-sample
spec: {}
EOF
```

3. Forward istio gateway port
```bash
kubectl port-forward -n istio-system service/istio-ingressgateway 8080:80 &
```



4. Install toystore service:
```bash
 kubectl apply -f - <<EOF
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: toystore
  labels:
    app: toystore
spec:
  selector:
    matchLabels:
      app: toystore
  template:
    metadata:
      labels:
        app: toystore
    spec:
      containers:
        - name: toystore
          image: quay.io/3scale/authorino:echo-api
          env:
            - name: PORT
              value: "3000"
          ports:
            - containerPort: 3000
              name: http
  replicas: 1
---
apiVersion: v1
kind: Service
metadata:
  name: toystore
spec:
  selector:
    app: toystore
  ports:
    - name: http
      port: 80
      protocol: TCP
      targetPort: 3000
EOF
```

5. Add a HTTPRoute for the toystore service
```bash
kubectl apply -f - <<EOF
---
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: HTTPRoute
metadata:
  name: toystore
  labels:
    app: toystore
spec:
  parentRefs:
    - name: istio-ingressgateway
      namespace: istio-system
  hostnames: ["*.toystore.com"]
  rules:
    - matches:
        - path:
            type: PathPrefix
            value: "/toy"
          method: GET
      backendRefs:
        - name: toystore
          port: 80
EOF
```

6. Check if the service is running
```bash
curl -v -H 'Host: api.toystore.com' http://localhost:8080/toy
```
_It should return a `200`_

7. Apply a RateLimit Policy
```bash
kubectl apply -f - <<EOF
---
apiVersion: kuadrant.io/v1beta1
kind: RateLimitPolicy
metadata:
  name: toystore
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
  rateLimits:
    - rules:
        - hosts: ["rate-limited.toystore.com"]
      configurations:
        - actions:
            - generic_key:
                descriptor_key: "limited"
                descriptor_value: "1"
      limits:
        - conditions:
            - "limited == 1"
          maxValue: 5
          seconds: 10
          variables: []
EOF
```

8. Now hitting the rate limited endpoint of toystore service should allow only 5 every 10 seconds. If you do 6 consecutive curls, the last one should return a `429`. It may take a while until the policy is applied.
```bash
while :; do curl --write-out '%{http_code}' --silent --output /dev/null -H "Host: rate-limited.toystore.com" -X GET http://localhost:8080/toy | egrep --color "\b(429)\b|$"; sleep 1; done
```

9. Profit!

## Steps
1. [x] Create a base branch for the merge. (This one)
2. [x] Bring the entire kuadrant-controller _git_ history, merge it in a sub-folder (28c815e67477b531b0fbcefb2742d0e571b8c451)
3. [x] Upgrade and match dependencies (https://github.com/Kuadrant/kuadrant-operator/pull/48/commits/84c936b46f580abfe1f56cade1f2d8ab5bb44d38)
4. [x] Source APIs (https://github.com/Kuadrant/kuadrant-operator/pull/50)
5. [x] Kubebuilder config files (https://github.com/Kuadrant/kuadrant-operator/pull/51)
6. [x] Fixing dependencies (https://github.com/Kuadrant/kuadrant-operator/pull/52)
7. [x] Controllers for reconciliation process (https://github.com/Kuadrant/kuadrant-operator/pull/53)
8. [x] Tests (https://github.com/Kuadrant/kuadrant-operator/pull/55)
9. [x] Documentation (https://github.com/Kuadrant/kuadrant-operator/pull/63)
10. [x] Remove leftovers (https://github.com/Kuadrant/kuadrant-operator/pull/64)
11. [x] Code Style tasks (https://github.com/Kuadrant/kuadrant-operator/pull/67)
12. [x] Fix User Guides and MeshConfig Bug (https://github.com/Kuadrant/kuadrant-operator/pull/70)
